tools/project_manager/diff.php

```diff
@@ -135,17 +138,44 @@
         $_GET["format"] = "remove";
     }
     $format_url = "?" . attr_safe(http_build_query($_GET));
+    $format_label = html_safe($format_label);
     echo "| <a href='$format_url'>$format_label</a>\n";
 }
 
+if (user_can_mentor_in_any_round()) {
+    // show option to change diff style
+    if ($bb_diffs) {
+        $bb_diffs_label = _("Show table-style diff");
+        $_GET["bb_diffs"] = "off";
+    } else {
+        $bb_diffs_label = _("Show BBCode-style diff");
+        $_GET["bb_diffs"] = "on";
+    }
+    $bb_diffs_url = "?" . attr_safe(http_build_query($_GET));
+    $bb_diffs_label = html_safe($bb_diffs_label);
+    echo "| <a href='$bb_diffs_url'>$bb_diffs_label</a>\n";
+}
+
+$diffurl = "$code_url/tools/project_manager/diff.php?project=$projectid&image=$image&L_round_num=$L_round_num&R_round_num=$R_round_num";
+
+// Provide button to copy BBCode diffs not wrapped in HTML
+if ($bb_diffs) {
+    $diffEngineCopy = new DifferenceEngineWrapperBB($diffurl);
+    $diffEngineCopy->setText($L_text, $R_text);
+    $cliptext = javascript_safe($diffEngineCopy->getDiff($L_label, $R_label));
+    echo "| \n<input type='button' value='" . attr_safe(_("Copy BBCode to clipboard")) . "' onClick='navigator.clipboard.writeText(\"$cliptext\");'>";
```

Comment:
`writeText` returns a promise while the user interacts with the permission dialog. If they reject the request or have turned off clipboard access altogether, that promise is rejected. We may want to in that case alert the user the copy failed.
```js
navigator.clipboard.writeText(\"$cliptext\").catch(function () {alert("Unable to copy!");});'
```
but that is getting a bit much. Maybe at this point we inline some js at the top with a click handler and just reference it here in the attribute.

Reply:
> @chrismiceli - I added
> 
> ```
>     "js_data" => "
>         function copyToClip(textstring) {
>             navigator.clipboard.writeText(textstring).catch(function () {alert('Unable to copy!');});
>         }
>     ",
> ```
> 
> to `$extra_args` and changed the onClick to
> 
> ```
> onClick='copyToClip(\"$cliptext\");'
> ```
> 
> Is that the kind of thing you had in mind? I am unable to reliably get my browser to reject use of clipboard, but have seen the "Unable to copy" alert. If you are able to control clipboard use easily on your system, this edit is in my [main sandbox](https://www.pgdp.org/~windymilla/c/) but not committed to the bbdiff branch yet if you're able to try it out, or suggest something different. Thanks

That's better. We'd need to encode the $cliptext though for the html attribute context to confirm no characters will escape out of the attribute like ".

Reply:
Doesn't `javascript_safe` already do that to the cliptext before it gets passed in as an argument to `copyToClip`? Or did you mean something else?

Reply:
javascript_safe is for encoding things to be used in a javascript context. Html attribute safe if for things going in html attributes. You are doing both here. This is why I was suggesting we pull the diff from the DOM to avoid all encoding issues.

Reply:
Ah yes. Well if the user feedback means we end up being able to use the DOM idea, that will clearly be better. It still might happen that we don't have the button at all. Depending on what others think, the first feedback I got was that they wouldn't want all the diffs in one block in the clipboard, just manually selecting one at a time. We'll see - thanks for continuing to educate me :)